Automattic/jetpack
# Search: ensure overlay trigger is consistently set

We currently have three ways to trigger the Instant Search overlay:
* When results are available (aka 'results')
* Immediately - as soon as the user interacts with the input (aka 'immediately')
* On form submit (aka 'submit')

https://github.com/Automattic/jetpack/issues/20849 notes that the "when results loaded" overlay trigger is being set as the default overlay trigger for new users. This isn't ideal, as it's one we intend to retire, and is perhaps the poorest user experience of the three options available ('immediate', 'results', and 'submit').

On digging into this further, I noticed that we sometimes specify 'immediate' as the default in the codebase, but we specify 'results' in the one spot where a new user's options are set.

#### Changes proposed in this Pull Request:
* Consolidate the default overlay trigger and always use a constant `DEFAULT_OVERLAY_TRIGGER` in the `Options` class.
* Set the default overlay trigger to 'submit', which fixes part of https://github.com/Automattic/jetpack/issues/23394.

##### Why 'submit'?

Opening on form submit is the least surprising and most accessible way for the modal to operate. When a user completes a search form, being navigated to the modal feels pretty natural. By setting this as the default, we're encouraging best practice. We shouldn't change the user context without their choice.

We worried for a while that Instant Search wouldn't feel 'instant' any more if we made this change. After a lot of thought, I think the important 'instant' is in the fantastically quick initial results, and in the lightning-fast sorting and filtering. The appearance of the modal halfway through typing a query is generally not a benefit.

The 'immediately' option will remain available in config for users that want it (along with the [special classes that trigger a modal in the docs](https://jetpack.com/jetpack_support_tag/search/)), and ultimately we'll aim to retire the 'results' option for the reasons outlined in #20849.

<img width="261" alt="Screen Shot 2022-07-14 at 15 22 36" src="https://user-images.githubusercontent.com/17325/178891241-426242a1-6ccf-4055-b1ad-2fd0d063f232.png">

#### Other information:


#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:
New site:
1. [Create a new test site](https://jurassic.ninja/create/?jetpack-beta&branches.jetpack-search=fix/consolidate-overlay-trigger-defaults&nojetpack) and purchase a Jetpack Search plan.
2. Ensure the front end of the site uses the 'submit' overlay trigger (i.e. the modal appears when you submit the search form or hit Enter).
3. Navigate to the Customberg UI and ensure that 'When user submits open' is set as the overlay trigger.

Respect existing setting:
1. [Create a new test site on trunk](https://jurassic.ninja/create/?jetpack-beta&branches.jetpack-search=trunk&nojetpack) and purchase a Jetpack Search plan.
2. Navigate to the Customberg UI and set the overlay trigger to 'Open when the user starts typing'.
3. Use Jetpack Beta to switch to the `fix/consolidate-overlay-trigger-defaults` feature branch.
4. Navigate to the Customberg UI and check that the overlay trigger is still set to 'Open when the user starts typing'.
5. Navigate to the front end of the site and ensure the overlay is triggered immediately when you start typing in a search input.
